File: LAB--2013-11-07/FileLab.java
import java.util.Scanner;
import java.io.File;
import java.io.PrintWriter;

public class FileLab {

    public static void main(String [] args) throws Exception {
        System.out.println("There are " + countK() + " 'k's in the first 10,000 words.");
        
        final int lineNum = 12345;
        System.out.println("The word on line " + lineNum + " is " + atLine(lineNum) + ".");
        
        System.out.println(wordsWithLength(2));
        System.out.println(wordsWithLength(3));
        System.out.println(wordsWithLength(4));
        System.out.println(wordsThatEndWith('Q'));
        System.out.println(aNotB('Q', 'U'));
        System.out.println(onlyAtIndex('B', 3));
    }
    
    public static int countK() throws Exception {
        Scanner words = new Scanner(new File("words.txt"));
        int kCount = 0;
        String currentWord;
        
        for (int i = 0; i < 10000; i++) {
            currentWord = words.next();
            if (currentWord.indexOf('K') != -1) {
                kCount++;
            }
            
        }    
        
        words.close();
        return kCount;
    }
    
    public static String atLine(int lineNumber) throws Exception {
        Scanner words = new Scanner(new File("words.txt"));
        String currentWord = "";
        
        for (int i = 0; i < lineNumber; i++) {
            currentWord = words.next();
        }
        
        return currentWord;
    }
    
    public static String wordsWithLength(int wordLength) throws Exception{
        Scanner words = new Scanner(new File("words.txt"));
        String currentWord;
        int wordCount = 0;
        
        while (words.hasNext()) {
            currentWord = words.next();
            
            if (currentWord.length() == wordLength) {
                wordCount++;
            }
            
        }
        
        words.close();
        return "There are " + wordCount + " words with " + wordLength + " letters.";
    }
    
    public static String wordsThatEndWith(char testChar) throws Exception {
        Scanner words = new Scanner(new File("words.txt"));
        String currentWord;
        int wordCount = 0;
        char lastChar = 'a';
        
        while (words.hasNext()) {
            currentWord = words.next();
            lastChar = currentWord.charAt(currentWord.length() - 1);
            
            
            if (lastChar == testChar) {
                wordCount++;
            }
            
        }
        
        words.close();
        return "There are " + wordCount + " words ending with char '" + testChar + "'.";
    }
    
    public static String aNotB(char a, char b) throws Exception {
        Scanner words = new Scanner(new File("words.txt"));
        int count = 0;
        String currentWord = "";
        
        while (words.hasNext()) {
            currentWord = words.next();
            boolean hasA = currentWord.indexOf(a) != -1;
            boolean hasB = currentWord.indexOf(b) != -1;
            
            if (hasA && !hasB) {
                count++;
            }
            
        }    
        
        words.close();
        return "There are " + count + " words that contain char '" + a + "' but not char '" + b + "'.";
    }
    
    public static String onlyAtIndex(char testChar, int testIndex) throws Exception {
        Scanner words = new Scanner(new File("words.txt"));
        int count = 0;
        String currentWord = "";
        boolean hasAtTestIndex = false;
        boolean hasAtZero = true;
        boolean hasAtOne = true;
        
        while (words.hasNext()) {
            currentWord = words.next();
            if (currentWord.length() - 1 > testIndex) {
            hasAtTestIndex = currentWord.charAt(testIndex) == testChar;
            hasAtZero = currentWord.charAt(0) == testChar;
            
            if(currentWord.length() > 1) {
                hasAtOne = currentWord.charAt(1) == testChar;
            }
            
            if (hasAtTestIndex && !hasAtZero && !hasAtOne) {
                count++;
            }
            }
        }    
        
        words.close();
        return "There are " + count + " words that have char '" + testChar + "' at index " + testIndex + ", but not 0 or 2.";
    }
}

File: LAB--2013-11-07/FileLab.py
def main():
    print "There are " + str(count_K()) + " 'k's in the first 10,000 words."

    line_num = 12345
    print "The word on line " + str(line_num) + " is " + at_line(line_num) + "."

    print words_with_length(2)
    print words_with_length(3)
    print words_with_length(4)
    print words_that_end_with("Q")
    print a_not_b("Q", "U")
    print only_at_index("B", 3)

def count_K():
    words = open("words.txt", "r")
    kCount = 0

    for i in range(0, 10000):
        current_word = words.readline()
        if current_word.find("K") != -1:
            kCount += 1

    words.close()
    return kCount

def at_line(line_number):
    words = open("words.txt", "r")

    for i in range(0, line_number):
        current_word = words.readline()

    words.close()
    return current_word.rstrip()

def words_with_length(word_length):
    words = open("words.txt", "r")
    word_count = 0

    for i in words:
        if len(i.rstrip()) == word_length:
            word_count += 1

    words.close()
    return "There are {0} words with {1} letters.".format(word_count, word_length)

def words_that_end_with(test_char):
    words = open("words.txt", "r")
    count = 0

    for i in words:
        last_char = i.rstrip()[-1:]
        
        if last_char == test_char:
            count += 1

    words.close()
    return "There are {0} words ending with char '{1}'.".format(count, test_char)

def a_not_b(a, b):
    words = open("words.txt", "r")
    count = 0

    for i in words:
        has_a = i.find(a) != -1
        has_b = i.find(b) != -1

        if has_a and not has_b:
            count += 1 

    words.close()
    return "There are {0} words that contain char '{1}' but not char '{2}'.".format(count, a, b)

def only_at_index(test_char, test_index):
    words = open("words.txt", "r")
    count = 0
    has_at_test_index = False
    has_at_one = True
    has_at_zero = True

    for i in words:
        if len(i) - 1 > test_index:
            has_at_test_index = i[test_index] == test_char
            has_at_zero = i[0] == test_char

        if len(i) > 1:
            has_at_one = i[1] == test_char

        meets_conditions = has_at_test_index and not has_at_zero and not has_at_one 

        if meets_conditions:
            count += 1

    words.close()
    return "There are {0} words that have char '{1}' at index {2}, but not at 0 or 2.".format(count, test_char, test_index)

if __name__ == "__main__": main()
